### PR TITLE
reduce kubelet getting node lease

### PR DESCRIPTION
**Pull request by answer1991**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

In a large-scale cluster, kubelet get lease too much and too frequently, which may cause etcd and apiserver performance issue. In an 8k nodes cluster, kubelet get lease 50k/min .

Node's lease only update by kubelet itself, so kubelet do not need to get lease every time when try to update, just use the latest one kubelet updated.

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

None

**Does this PR introduce a user-facing change?**:

```release-note
Fix kubelet NodeLease potential performance issues. Kubelet now will try to update lease using cached one instead of get from API Server every time.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

None

### CODE REVIEW

File: pkg/kubelet/nodelease/controller_test.go [hunk 1]
```diff
@@ -282,3 +282,111 @@ func TestRetryUpdateLease(t *testing.T) {
...
+
+func TestUpdateUsingLatestLease(t *testing.T) {
+	node := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+			UID:  types.UID("foo-uid"),
+		},
+	}
+
+	gr := schema.GroupResource{Group: "v1", Resource: "lease"}
```
Comment (wojtek-t) [inline pkg/kubelet/nodelease/controller_test.go]: Group isn't actually v1.
```
import "k8s.io/kubernetes/pkg/apis/coordination"

notFoundErr := apierrors.NewNotFound(coordination.Resource("lease"), "foo")
```
coordination.Resource("lease")

File: pkg/kubelet/nodelease/controller_test.go [hunk 2]
```diff
@@ -282,3 +282,111 @@ func TestRetryUpdateLease(t *testing.T) {
...
+	notFoundErr := apierrors.NewNotFound(gr, "foo")
+	cases := []struct {
+		desc                       string
+		latestLease                *coordinationv1.Lease
+		updateReactor              func(action clienttesting.Action) (bool, runtime.Object, error)
+		getReactor                 func(action clienttesting.Action) (bool, runtime.Object, error)
+		createReactor              func(action clienttesting.Action) (bool, runtime.Object, error)
+		onRepeatedHeartbeatFailure func()
```
Comment (wojtek-t) [inline pkg/kubelet/nodelease/controller_test.go]: it seems to be nil in all cases - please remove it

File: pkg/kubelet/nodelease/controller_test.go [hunk 3]
```diff
@@ -282,3 +282,111 @@ func TestRetryUpdateLease(t *testing.T) {
...
+	}{
+		{
+			desc:          "latestLease is nil and need to create",
+			latestLease:   nil,
+			updateReactor: nil,
+			getReactor: func(action clienttesting.Action) (bool, runtime.Object, error) {
+				return true, nil, notFoundErr
+			},
+			createReactor: func(action clienttesting.Action) (b bool, object runtime.Object, e error) {
+				return true, &coordinationv1.Lease{ObjectMeta: metav1.ObjectMeta{ResourceVersion: "1"}}, nil
```
Comment (wojtek-t) [inline pkg/kubelet/nodelease/controller_test.go]: Currently it doesn't change much, but I think we should set at least name and namespace.

Sp let's introduce a simple helper function
`makeLease(name, resourceVersion) *coordinationv1.Lease`
and use it here and below

File: pkg/kubelet/nodelease/controller_test.go [hunk 4]
```diff
@@ -282,3 +282,111 @@ func TestRetryUpdateLease(t *testing.T) {
...
+			updateReactor: func(action clienttesting.Action) (bool, runtime.Object, error) {
+				return true, &coordinationv1.Lease{ObjectMeta: metav1.ObjectMeta{ResourceVersion: "2"}}, nil
+			},
+			onRepeatedHeartbeatFailure: nil,
+			expectLeaseResourceVersion: "2",
+		},
+		{
+			desc:        "update with latest lease failed",
+			latestLease: &coordinationv1.Lease{ObjectMeta: metav1.ObjectMeta{ResourceVersion: "1"}},
+			updateReactor: func() func(action clienttesting.Action) (b bool, object runtime.Object, e error) {
```
Comment (wojtek-t) [inline pkg/kubelet/nodelease/controller_test.go]: remove names of output params, i.e.

func (action clienttesting.Action) (bool, runtime.Object, error) {

File: pkg/kubelet/nodelease/controller_test.go [hunk 5]
```diff
@@ -282,3 +282,111 @@ func TestRetryUpdateLease(t *testing.T) {
...
+				i := 0
+				return func(action clienttesting.Action) (bool, runtime.Object, error) {
+					i++
+					switch i {
+					case 1:
+						return true, nil, notFoundErr
+					case 2:
+						return true, &coordinationv1.Lease{ObjectMeta: metav1.ObjectMeta{ResourceVersion: "3"}}, nil
+					default:
+						return true, &coordinationv1.Lease{}, nil
```
Comment (wojtek-t) [inline pkg/kubelet/nodelease/controller_test.go]: this should never happen - so instead of failing silently, it's better to do t.Fatalf(....)

**Review by wojtek-t — COMMENTED** [review]
I added couple comments, but in general I really like this.

File: pkg/kubelet/nodelease/controller.go [hunk 6]
```diff
@@ -87,7 +88,18 @@ func (c *controller) Run(stopCh <-chan struct{}) {
 }
 
 func (c *controller) sync() {
+	if c.latestLease != nil {
+		lease, err := c.leaseClient.Update(c.newLease(c.latestLease))
```
[the remark this answers is not in the document]
Comment (answer1991) [reply pkg/kubelet/nodelease/controller.go]: thanks, done

File: pkg/kubelet/nodelease/controller_test.go (hunk 2 above)
Comment (answer1991) [reply pkg/kubelet/nodelease/controller_test.go]: done

File: pkg/kubelet/nodelease/controller_test.go (hunk 3 above)
Comment (answer1991) [reply pkg/kubelet/nodelease/controller_test.go]: done

File: pkg/kubelet/nodelease/controller_test.go (hunk 5 above)
Comment (answer1991) [reply pkg/kubelet/nodelease/controller_test.go]: done

File: pkg/kubelet/nodelease/controller_test.go (hunk 1 above)
Comment (answer1991) [reply pkg/kubelet/nodelease/controller_test.go]: should be `coordinationv1`?

```go
import coordinationv1 "k8s.io/api/coordination/v1"

apierrors.NewNotFound(coordinationv1.Resource("lease"), "foo")
```

**Comment by wojtek-t** [comment]
LGTM - but I will let others to also take a look before approving.

File: pkg/kubelet/nodelease/controller_test.go (hunk 1 above)
Comment (wojtek-t) [reply pkg/kubelet/nodelease/controller_test.go]: in this context both are fine actually (because version is dropped for group resource)

Comment (answer1991) [reply pkg/kubelet/nodelease/controller_test.go]: ok, thanks

**Pull request merged**